core/src/test/java/org/apache/iceberg/io/TestAppenderFactory.java

```diff
@@ -0,0 +1,265 @@
+        table.encryption(), 1, 1);
+  }
+
+  private PartitionKey createPartitionKey() {
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+
+    Record record = GenericRecord.create(table.spec().schema()).copy(ImmutableMap.of("data", "aaa"));
+
+    PartitionKey partitionKey = new PartitionKey(table.spec(), table.schema());
+    partitionKey.partition(record);
+
+    return partitionKey;
+  }
+
+  private List<T> testRowSet() {
+    return Lists.newArrayList(
+        createRow(1, "aaa"),
+        createRow(2, "bbb"),
+        createRow(3, "ccc"),
+        createRow(4, "ddd"),
+        createRow(5, "eee")
+    );
+  }
+
+  private DataFile prepareDataFile(List<T> rowSet, FileAppenderFactory<T> appenderFactory,
+                                   OutputFileFactory outputFileFactory) throws IOException {
+    DataWriter<T> writer = appenderFactory.newDataWriter(outputFileFactory.newOutputFile(), format, partition);
+    try (DataWriter<T> closeableWriter = writer) {
+      for (T row : rowSet) {
+        closeableWriter.add(row);
+      }
+    }
+
+    return writer.toDataFile();
+  }
+
+  @Test
+  public void testDataWriter() throws IOException {
+    FileAppenderFactory<T> appenderFactory = createAppenderFactory(null, null, null);
+    OutputFileFactory outputFileFactory = createFileFactory();
+
+    List<T> rowSet = testRowSet();
+    DataFile dataFile = prepareDataFile(rowSet, appenderFactory, outputFileFactory);
+
+    table.newRowDelta()
+        .addRows(dataFile)
+        .commit();
+
+    Assert.assertEquals("Should have the expected records.", expectedRowSet(rowSet), actualRowSet("*"));
+  }
+
+  @Test
+  public void testEqDeleteWriter() throws IOException {
+    List<Integer> equalityFieldIds = Lists.newArrayList(table.schema().findField("id").fieldId());
+    FileAppenderFactory<T> appenderFactory = createAppenderFactory(equalityFieldIds,
+        table.schema().select("id"), null);
+    OutputFileFactory outputFileFactory = createFileFactory();
+
+    List<T> rowSet = testRowSet();
+    DataFile dataFile = prepareDataFile(rowSet, appenderFactory, outputFileFactory);
+
+    table.newRowDelta()
+        .addRows(dataFile)
+        .commit();
+
+    List<T> deletes = Lists.newArrayList(
+        createRow(1, "aaa"),
+        createRow(3, "bbb"),
+        createRow(5, "ccc")
+    );
+    EqualityDeleteWriter<T> eqDeleteWriter =
+        appenderFactory.newEqDeleteWriter(outputFileFactory.newOutputFile(), format, partition);
+    try (EqualityDeleteWriter<T> closeableWriter = eqDeleteWriter) {
+      closeableWriter.deleteAll(deletes);
+    }
+
+    table.newRowDelta()
+        .addDeletes(eqDeleteWriter.toDeleteFile())
+        .commit();
+
+    List<T> expected = Lists.newArrayList(
+        createRow(2, "bbb"),
+        createRow(4, "ddd")
+    );
+    Assert.assertEquals("Should have the expected records", expectedRowSet(expected), actualRowSet("*"));
```

Comment:
Instead of reading from the table, I would rather see a test that the equality delete file contains the expected row data. In this case, it should not contain the `data` column. I would like to see that checked. And it would be good to add a case where the whole original row is written to the file.